imap provider: attachments / headers / content-section parsing.

benrussell opened the pull request:
# 🗣 Description

Extend imap provider with MIME parsing for attachments and content-sections.

resolves #4605 - "List email attachment filenames"

resolves #4791 ? - HTML/Text content sections are extracted

resolves #4792 - Add support for parsing email headers.

resolves #4793 - parse attachments blobs


## 🔨 Related Issues




## 📄 Documentation Updates

Adds new fields to the schema


## 🤔 Concerns

Uses the mail-parser crate: https://crates.io/crates/mail-parser
Found a bug with the attachment name decoding for the upstream crate mail-parser.
Will need to patch and submit for upstream crate.
Added a unit test that relies on the broken behaviour for now.

MIME attachment rules allow for an edge case where an attachment does not require either a mime_type field OR a filename.
Have dealt with this by assigning the default mime type of "application/octet-stream".
This mime-type is used to prevent a user-agent from attempting to display binary data as text to the end user.

benrussell committed:
imap provider: Adds attachment parsing module.

peasee commented on crates/data_components/src/imap/attachments.rs:
```diff
@@ -0,0 +1,367 @@
+                        }
+                    )
+                },
+                (None, Some(t)) => {
+                    //no filename, mime type only
+                    ret.push(
+                        AttachmentInfo{
+                            filename: None,
+                            mime_type: Some(t.clone()),
+                        }
+                    )
+                },
+                (None,None) => {
+                    //panic!("no filename and no mime type.");
+                    //FIXME: This is a legal MIME state. How to handle for Spice?
```
I think for the case where there is no filename, regardless of MIME type, we could introduce a new connector parameter to control whether we log a warning or error like `invalid_attachment_action: warn`.

It could accept values similar to the existing [`unsupported_type_action`](https://spiceai.org/docs/reference/spicepod/datasets#unsupported_type_action) (`warn`, `error`, `ignore`), but it would be a [connector parameter](https://github.com/spiceai/spiceai/blob/trunk/crates/runtime/src/dataconnector/imap.rs#L41) instead of a dataset parameter and behavior would be basically the same.

I think for most cases a warning would suffice, but allowing an error here could be useful.

benrussell replied:
Have added notes to review the Email MIME RFC. 

IIRC filename= param for attachments is a "suggested default" for mail clients.

The RFC may also define that bare attachment records such as this are plain text. 

If the above is the case perhaps a default to untitled.txt is a useful path.


Connector param as suggested may be the best path for a Spice provider.

benrussell replied:
Defaulted mime_type to application/octet-stream.

Have not added connector param for raising warning.

Have left marker in code so it can be addressed if it's still wanted.

peasee commented on crates/data_components/src/imap/attachments.rs:
```diff
@@ -0,0 +1,367 @@
+
+            } //loop attachments for email
+
+        } //can we parse this email?
```
this could be a good place to add an `else {}` and handle based on the `invalid_attachment_action`

benrussell replied:
Code LOC affected has changed with refactor. Tagged with FIXME's so it's easy to find in latest patch.

peasee commented on crates/data_components/src/imap/provider.rs:
```diff
@@ -112,6 +116,21 @@ impl TableProvider for ImapTableProvider {
         if self.fetch_content {
             fields.push(Field::new("content", DataType::Utf8, true));
+
+            fields.push(
+                Field::new(
+                "attachment_name", //FIXME: what to use as field name? attachments.rs uses filename
+                DataType::List(Arc::new(Field::new("item", DataType::Utf8, true))),
+            true)
+            );
+
+            fields.push(
+                Field::new(
+                    "attachment_mime_type",
+                    DataType::List(Arc::new(Field::new("item", DataType::Utf8, true))),
+                    true)
+            );
```
I think we should use a struct field here, like `attachments`. This also avoids us needing to keep separate vecs for filenames and mime type, so should help simplify some of the parsing where we compare filename vs mime vec length.

E.g.

```
Field::new(
           "attachments",
           DataType::List(Arc::new(Field::new_list_field(
               DataType::Struct(Fields::from(vec![
                   Field::new("filename", DataType::Utf8, true),
                   Field::new("mime_type", DataType::Utf8, true),
               ])),
               true,
           ))),
           true,
       )
```

benrussell replied:
Nice. I wasn't aware that this is possible in Arrow.

I have refactored appropriately and created a new `fn build_listarray_for_attachments(...)` that hopefully does the right thing.

I'm not up to speed with creating Arrow data structs so I've tagged the new fn with a FIXME and appropriate notes.

peasee commented on crates/data_components/src/imap/attachments.rs:
```diff
@@ -0,0 +1,367 @@
+
+#[derive(Debug)]
+#[derive(PartialEq)]
+pub struct AttachmentRecords{
+    pub mime_types: Vec<Option<String>>,
+    pub filenames: Vec<Option<String>>,
+}
```
if we take a struct as our field, this gets simpler

```
pub struct AttachmentRecords {
    pub attachments: Vec<AttachmentInfo>,
}
```

benrussell replied:
I had started with this but split it into the two vectors and a tuple integration for easier integration with the existing Arrow examples.

As you can appreciate, it took extra work to zip() the two Vecs back together in the AttachmentRecords.collate() fn which is now dead and gone.

All good.

peasee commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -125,6 +131,12 @@ impl ImapTableProvider {
         if self.fetch_content {
             fields.push(Arc::new(StringArray::from(bodies)));
+
+            //FIXME: not sure if this is the correct place to add our records to the fields vec
```
💯 I think we would put attachment bodies here, but because it's just metadata at the moment IMO we could include it all the time. Even when we're getting attachment bodies, I think we should still support attachment metadata when `self.fetch_content == false`.

benrussell replied:
Have made some changes to suit this feedback also.

Have left a note as I don't believe the bodies data is available without a fetch_content (at least it's not as Rust vars) so that will need some consideration.

peasee reviewed (commented):
Thanks for working on this, I'm liking it! A couple comments around error handling and the field type(s).

benrussell committed:
imap provider: refac attachments mod to use AttachmentRecords struct containing single Vec<AttachmentInfo> instead of two Vec<String>.

benrussell committed:
imap provider: refac to use structured field (Vec<AttachmentInfo>) for attachments

benrussell committed:
imap provider: Adds support to optionally decode attachments into Vec<u8>. No schema/arrow field yet. #4793

benrussell committed:
imap provider: Improves attachment blob handling #4793

benrussell committed:
imap provider: Improves attachment blob handling #4793

benrussell commented:
Added code to parse and optionally store attachment blobs as decoded files.

benrussell committed:
prov/imap: refac AttachmentRecords::try_from(). No longer trait impl. Allows accept two args for no copy of email payload.

benrussell committed:
prov/imap: attachments.rs: refac try_from() to take a pre-parsed &Messages obj

benrussell committed:
prov/imap: created body_content.rs to filter html/text content-body sections for easier use.

benrussell committed:
prov/imap: Adds code to extract and store email headers

benrussell committed:
prov/imap: Adds code to extract and store email headers.

Improved test_data files.

Moved unit tests into tests.rs

benrussell committed:
prov/imap: Tidy code to gather data from parsed mime msg.

benrussell committed:
prov/imap: refac mime-parse code to use a ret struct instead of a tuple.

Seems like a trivial change, but it condenses the destructuring code into one small place where all LOC can be read and understood at once.

Previous commit uses a fragile tuple ret.
 Expected order of args in tuple was seperated from the return of the tuple data by a block of code.

benrussell committed:
prov/imap: imap query msg tied to self.fetch_content value

benrussell committed:
prov/imap: Condense mime msg parse extraction/wrap code.

benrussell committed:
prov/imap: Fix new imap query string.

benrussell committed:
prov/imap: Moved+renamed schema fields for body_content_records.

Now only defined if self.fetch_content

Name changed to content_sections

benrussell committed:
prov/imap: minor: tidy attachments.rs

benrussell committed:
prov/imap: minor: #[cfg(test)] filter for tests.rs

benrussell committed:
prov/imap: Tweaks to match schema() and build_record_batch() data structures.

benrussell committed:
prov/imap: Extract header data directly from imap Fetch handle.

benrussell committed:
prov/imap: Rename body_content.rs structs for clarity of purpose.

benrussell committed:
prov/imap: minor: tweaked notes

benrussell committed:
prov/imap: removes store_blobs option from attachment parser. Now always stored.

benrussell committed:
prov/imap: Condense parsed-mime extraction.

benrussell committed:
prov/imap: Added content_sections builder fn for build_recordbatch()

benrussell committed:
prov/imap: refac: mv attachments.rs -> mime.rs

benrussell committed:
prov/imap: refac: merge body_content.rs into mime.rs. Adds MimeExtract wrapper to make integration with provider.rs cleaner.

benrussell committed:
prov/imap: mime.rs: promoted various ..try_from() fns to proper trait impls.

benrussell committed:
prov/imap: Consolidate some FIXME notes.

benrussell committed:
prov/imap: minor: tidy and docs

benrussell committed:
prov/imap: Attachments that decode to empty vecs are now repr as None

benrussell committed:
prov/imap: Attachments with no filename or mime_type now default to "application/octet-stream"

Added note for RFC 2046 which is MIME pt 2

Defaulted mime_type to app/octet-stream to stop the potential display of binary data as text to the user.

Added note about maybe storing the charset of attachments. The RFC mentions it for text/* segments.

peasee commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -74,6 +82,105 @@ fn build_listarray_for_strings(values: Vec<Option<Vec<Option<String>>>>) -> List
+            for attachment in attachments {
+                let struct_builder = list_builder.values();
+
+                //FIXME: the expect() msgs in this code seem like they could be improved? (attachments)
```
Yes, we typically don't merge code with `.expect()` unless used in tests. We should make some new errors and wrap these instead with `.context()`. The signature of this function would need to change to `Result<ListArray, Error>`

benrussell replied:
precedent:

https://github.com/spiceai/spiceai/blob/trunk/crates/db_connection_pool/src/dbconnection/snowflakeconn.rs#L708

https://github.com/spiceai/spiceai/blob/trunk/crates/db_connection_pool/src/dbconnection/snowflakeconn.rs#L739


I went looking for existing Spice OSS code to create a suitable Arrow Structure record type when refac to compound attachment struct records.

Found the snowflakeconn.rs example.

Would not normally submit with loose code paths.

fn's with .expect in PR have been FIXME tagged for further review since creation as I know they're a pain point.

phillipleblanc replied:
As part of our style guide, we don't allow the usage of `.expect` or `.unwrap` in production code - we allow `.expect` only in tests. Those two examples you found are helper methods in a test module. A more idiomatic example you can reference would be: https://github.com/spiceai/spiceai/blob/trunk/crates/db_connection_pool/src/dbconnection/snowflakeconn.rs#L257

benrussell replied:
Fixed.

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+        let attachment_records = AttachmentRecords::try_from( &message ).unwrap();
+
+        let extracted_set = attachment_records.attachments;
+        // for rec in &collated{
+        //     println!("-rec: {:?}", rec);
+        // }
```
```suggestion
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+    let eml_msg = MessageParser::default().parse(&raw_email);
+
+    if let Some(message) = eml_msg {
+        //println!("message html_body_count: {}", message.html_body_count());
+        let _content_records = ContentSectionRecords::try_from( &message ).unwrap();
+        //println!("\ncontent_records = {:#?}", content_records);
```
```suggestion
        let _content_records = ContentSectionRecords::try_from( &message ).expect("Failed to parse message");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+    let eml_msg = MessageParser::default().parse(&raw_email);
+
+    if let Some(message) = eml_msg {
+        //println!("message html_body_count: {}", message.html_body_count());
+        let _content_records = ContentSectionRecords::try_from( &message ).unwrap();
+        //println!("\ncontent_records = {:#?}", content_records);
```
```suggestion
        let _content_records = ContentSectionRecords::try_from(&message).expect("Failed to parse message");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+    let eml_msg = MessageParser::default().parse(&raw_email);
+
+    if let Some(message) = eml_msg {
+        //println!("message html_body_count: {}", message.html_body_count());
+        let _content_records = ContentSectionRecords::try_from( &message ).unwrap();
+        //println!("\ncontent_records = {:#?}", content_records);
```
```suggestion
        let _content_records = ContentSectionRecords::try_from( &message ).expect("Failed to parse message");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+            let blob_str = if let Ok(blob) = String::from_utf8(blob) {
+                blob
+            } else {
+                format!("[{}] Could not decode into utf8 for display", rec.mime_type.clone().unwrap())
```
```suggestion
                format!("[{}] Could not decode into utf8 for display", rec.mime_type.clone().expect("mime type not set"))
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+        }
+
+
+        //let image_bytes = include_bytes!("test_data/Mark - Orange on White.png");
```
```suggestion
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+    let eml_msg = MessageParser::default().parse(&raw_email);
+
+    if let Some(message) = eml_msg {
+        //println!("message html_body_count: {}", message.html_body_count());
+        let _content_records = ContentSectionRecords::try_from( &message ).unwrap();
+        //println!("\ncontent_records = {:#?}", content_records);
```
```suggestion
        let _content_records = ContentSectionRecords::try_from(&message).expect("Failed to parse message");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+        //println!("\ncontent_records = {:#?}", content_records);
+
+
+        let attachment_records = AttachmentRecords::try_from( &message ).unwrap();
```
```suggestion
        let attachment_records = AttachmentRecords::try_from( &message ).expect("Failed to parse attachments");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+        //println!("\ncontent_records = {:#?}", content_records);
+
+
+        let attachment_records = AttachmentRecords::try_from(&message).unwrap();
```
```suggestion
        let attachment_records = AttachmentRecords::try_from(&message).expect("Failed to parse attachments");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+
+        let extracted_set = attachment_records.attachments;
+        for rec in &extracted_set {
+            let blob = rec.blob.clone().unwrap();
```
```suggestion
            let blob = rec.blob.clone().expect("blob not set");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+        //println!("\ncontent_records = {:#?}", content_records);
+
+
+        let attachment_records = AttachmentRecords::try_from(&message).unwrap();
```
```suggestion
        let attachment_records = AttachmentRecords::try_from(&message).expect("Failed to parse attachments");
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+        let attachment_records = AttachmentRecords::try_from(&message).unwrap();
+
+        let extracted_set = attachment_records.attachments;
+        // for rec in &collated{
+        //     println!("-rec: {:?}", rec);
+        // }
```
```suggestion
```

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+                mime_type: Some("application/pdf".to_string()),
+                blob: None
+            },
+            // AttachmentInfo{
+            //     filename: None,
+            //     mime_type: None,
+            // },
```
```suggestion
```

benrussell replied:
Updated this to reflect the new default mime_type assignment of "application/octet-stream".

This triggered a fail on my cargo test but I didn't see it before I'd pushed and requested review.

phillipleblanc commented on crates/data_components/src/imap/provider.rs:
```diff
@@ -152,26 +198,32 @@ impl TableProvider for ImapTableProvider {
             };
 
+
+            let mime_extract = MimeExtract::try_from(&body).ok().unwrap();
```
This will need to return a DataFusion error instead of unwrapping.

benrussell replied:
MimeExtract can always return a valid struct. 
Worst case is None vals for member vars. 
Removes need for DataFusion error wrap.

Have refactored to use MimeExtract::from(...)

https://github.com/spiceai/spiceai/pull/5083/commits/a44621a75b7d3a18bca32f305323e59841845841

phillipleblanc commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -74,6 +82,105 @@ fn build_listarray_for_strings(values: Vec<Option<Vec<Option<String>>>>) -> List
 }
 
+
+//FIXME: The following fn were written with help from GPT.
+// They seem plausible but I am not versed in the construction of Arrow suitable structures.
+// I had a look at snowflakeconn.rs - seems similar.
+// - build_listarray_for_content_sections(..)
+// - build_listarray_for_attachments(..)
+
+
+
```
```suggestion
```

Thanks for the note!

phillipleblanc commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -74,6 +82,105 @@ fn build_listarray_for_strings(values: Vec<Option<Vec<Option<String>>>>) -> List
+
+
+
+fn build_listarray_for_content_sections(values: Vec<Option<Vec<ContentSectionInfo>>>) -> ListArray {
```
This function should return a Result<ListArray> and instead of panicking below, return an error that is handled higher up.

phillipleblanc replied:
Asking GPT to create some tests for this code would be useful as well.

benrussell replied:
return type, data, and handling has been upgraded.

benrussell replied:
Created tests with o3-mini-high - found and fixed a bug with the Arrow struct def code.

phillipleblanc commented on crates/data_components/src/imap/mime.rs:
```diff
@@ -0,0 +1,244 @@
+
+
+impl TryFrom<&Message<'_>> for AttachmentRecords{
+    type Error = String;
```
We should define a new Snafu Error enum and use that here.

benrussell replied:
Done.

https://github.com/spiceai/spiceai/pull/5083/commits/e15738f8dfd374adda1e45681206afbeaf48bada

phillipleblanc commented on crates/data_components/src/imap/mime.rs:
```diff
@@ -0,0 +1,244 @@
+
+
+impl TryFrom<&Message<'_>> for ContentSectionRecords {
+    type Error = String;
```
We should define a new Snafu Error enum and use that here.

benrussell replied:
Done.

https://github.com/spiceai/spiceai/pull/5083/commits/e15738f8dfd374adda1e45681206afbeaf48bada

phillipleblanc commented on crates/data_components/src/imap/tests.rs:
```diff
@@ -0,0 +1,221 @@
+
+
+#[test]
+fn eml_utf8() {
```
I would move these tests to the bottom of the mime.rs file instead of having them in their own file.

benrussell replied:
merged tests.rs into mime.rs

prefixed test function signatures with "test_"

benrussell committed:
Update crates/data_components/src/imap/tests.rs

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
Update crates/data_components/src/imap/mod.rs

removes GPT-assist noe

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better debug msg if eml_utf8 cotent records fail to parse.

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better dbg msg: eml_no_attachments parse content records

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better dbg msg: eml_three_attachments parse content records

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: debug msg if we try and log info from a None  mime_type record

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better debug msg if eml_malformed cotent records fail to parse.

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better debug msg if eml_utf8 attachments records fail to parse.

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better debug msg if eml_three_attachments attachments records fail to parse.

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better debug msg if eml_three_attachments cannot access blob for debug output to log

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: better debug msgs if we get parsing failures instead of just unwrap. cruft trim.

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: tests.rs: Sets mime_type to app/octet-stream if can't detect from MIME message and there's also no filename.

benrussell committed:
Merge remote-tracking branch 'github_br/imap_attachment_parsing' into imap_attachment_parsing

benrussell committed:
prov/imap: tests.rs: removes filename for local test png

Co-authored-by: Phillip LeBlanc <phillip@leblanc.tech>

benrussell committed:
prov/imap: Added some notes about fixes required.

benrussell committed:
prov/imap: Clippy lint easy stuff.

benrussell committed:
prov/imap: MimeExtract can always ret a valid struct. Worst case is None vals for member vars. Removes need for DataFusion error wrap.

benrussell committed:
prov/imap: MimeExtract: deletes try_from impl

benrussell committed:
prov/imap: mime.rs: Upgrade to snafu errors. Tweak From trait impl. Clippy liny.

benrussell committed:
prov/imap: Upgrade build_listarray_for_attachments(...) to use snafu errors and elim use of .expect()

benrussell committed:
prov/imap: Upgrade build_listarray_for_content_sections(...) to use snafu errors and elim use of .expect()

benrussell committed:
prov/imap: Upgrade build_recordbatch(..); better ArrowError handling

benrussell committed:
prov/imap: Merged tests.rs into mime.rs

prefixed existing test function signatures with "test_"

benrussell committed:
prov/imap: mime.rs: Improved mod{} wrap of test fns

benrussell committed:
prov/imap: mod.rs: Created a basic test fn for build_listarray_for_content_sections() fn using GPT. Compiles.

benrussell committed:
prov/imap: mod.rs: Created a basic test fn for build_listarray_for_attachments() fn using GPT. Immediately found a bug where the builder was not init correctly with the new BinaryBuilder field.

benrussell commented:
Clippy passes.

benrussell committed:
Merge branch 'spiceai:trunk' into imap_attachment_parsing

benrussell committed:
prov/imap: Fixes clippy lint re: header reconstruction in test case

benrussell committed:
prov/imap: Fixes clippy lint re: header reconstruction in test case

benrussell committed:
prov/imap: provider.rs: Tidy up to pass "make lint" rules.

benrussell committed:
prov/imap: mod.rs: Tidy up to pass "make lint" rules.

benrussell committed:
prov/imap: mime.rs: Tidy up to pass "make lint" rules.

benrussell committed:
prov/imap: Tidy mod.rs for unused imports.

benrussell committed:
prov/imap: Upgrade unit test to more modern idiom because clippy.

benrussell commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -57,6 +66,19 @@ pub enum Error {
     #[snafu(display("An invalid authentication mode was provided: {auth_mode}"))]
     InvalidAuthMode { auth_mode: String },
+
+    //FIXME: Not happy with error descriptions.
```
Are these error descriptions adequate / correct?

benrussell commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -74,6 +96,113 @@ fn build_listarray_for_strings(values: Vec<Option<Vec<Option<String>>>>) -> List
     builder.finish()
 }
 
+//FIXME: I don't really like the overload/shadow of Result here but GPT told me it was an accepted Rust idiom.
```
Please clarify which Error idiom is prefered in the Spice codebase so I can make it consistent. (1 of 2)

benrussell commented on crates/data_components/src/imap/mime.rs:
```diff
@@ -0,0 +1,413 @@
+    }
+}
+
+//FIXME: Apparently overload/shadow of "Error" is normal Rust idiomatic code.
```
Please clarify which Error idiom is prefered in the Spice codebase so I can make it consistent. (2 of 2)

benrussell committed:
prov/imap: minor: tidy deprecated code cruft

benrussell commented on crates/data_components/src/imap/mime.rs:
```diff
@@ -0,0 +1,412 @@
+    }
+}
+
+//FIXME: Apparently overload/shadow of "Error" is normal Rust idiomatic code.
```
Please clarify which Error idiom is prefered in the Spice codebase so I can make it consistent. (2 of 2)

benrussell commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -57,6 +66,19 @@ pub enum Error {
     InvalidSSLMode { ssl_mode: String },
     #[snafu(display("An invalid authentication mode was provided: {auth_mode}"))]
     InvalidAuthMode { auth_mode: String },
+
+    //FIXME: Not happy with error descriptions.
+    #[snafu(display("Should return a field builder: {field_number}/{field_name}"))]
+    FailedToBuildListArrayForAttachments {
+        field_number: i32,
+        field_name: String,
+    },
+
+    #[snafu(display("Should return a field builder: {field_number}/{field_name}"))]
+    FailedToBuildListArrayForContentSections {
+        field_number: i32,
+        field_name: String,
+    },
 }
```
Are these error descriptions adequate / correct?

benrussell commented on crates/data_components/src/imap/mod.rs:
```diff
@@ -74,6 +96,113 @@ fn build_listarray_for_strings(values: Vec<Option<Vec<Option<String>>>>) -> List
     builder.finish()
 }
 
+//FIXME: I don't really like the overload/shadow of Result here but GPT told me it was an accepted Rust idiom.
+pub type Result<T, E = Error> = std::result::Result<T, E>;
+
```
Please clarify which Error idiom is prefered in the Spice codebase so I can make it consistent. (1 of 2)

benrussell commented on crates/data_components/src/imap/mime.rs:
```diff
@@ -0,0 +1,412 @@
+            })
+    }
+}
+
+//FIXME: Apparently overload/shadow of "Error" is normal Rust idiomatic code.
+// not sure, went with this for now.
+#[derive(Debug, Snafu)]
+pub enum MimeExtractError {
+    #[snafu(display("No attachments found."))]
+    AttachmentsNotFound,
+
+    #[snafu(display("No content sections found."))]
+    ContentSectionsNotFound,
+}
+
+// Attachment extraction code.
+// filename, mime-type, blob
+
+#[derive(Debug, PartialEq)]
+pub struct AttachmentInfo {
+    pub mime_type: Option<String>,
+    pub filename: Option<String>,
+    pub blob: Option<Vec<u8>>,
+    //FIXME: store charset?
+}
+
+#[derive(Debug, PartialEq)]
+pub struct AttachmentRecords {
+    pub attachments: Vec<AttachmentInfo>,
+}
+
+impl TryFrom<&Message<'_>> for AttachmentRecords {
+    type Error = MimeExtractError;
+
```
Please clarify which Error idiom is prefered in the Spice codebase so I can make it consistent. (2 of 2)
